lib/rspec/support/method_signature_verifier.rb

```diff
@@ -85,12 +85,58 @@ def has_kw_args_in?(args)
             (args.last.empty? || args.last.keys.any? { |x| x.is_a?(Symbol) })
         end
 
-        # Without considering what the last arg is, could it
-        # contain keyword arguments?
         def could_contain_kw_args?(args)
+          # If the arg count is less than the required positional args there are no keyword args
           return false if args.count <= min_non_kw_args
 
-          @allows_any_kw_args || @allowed_kw_args.any?
+          # If there is no keyword splat or allowed keywords there are no keyword args
+          return false unless @allows_any_kw_args || @allowed_kw_args.any?
+
+          # If the args captured are more than the positional arguments then there are keywords
+          return true if args.count > max_non_kw_args
+
+          # Otherwise we need to check wether the last argument is a positional non default hash,
+          # e.g. that this:
+          #
+          # def capture(_, last = {}, keyword: 2)
+          #
+          # Is valid with both:
+          #
+          # [1, {:some => :hash}]
+          # [1, {:keyword => :value}]
+          #
+          # but for different reasons.
+
+          # If the keyword splat is allowed, and there are no specified
+          # keywords args then we will either have returned true on L100
+          # or this is invalid.
+          return false if @allowed_kw_args.empty? && @allows_any_kw_args
+
+          if Hash === args.last
+            last_arg_matches_any_keywords = !(args.last.keys & @allowed_kw_args).empty?
+            has_extra_keys = !(args.last.keys - @allowed_kw_args).empty?
+
+            # If we allow the keyword splat, and the last hash contains
+            # a specified keyword, then this hash contains keywords
+            return true if @allows_any_kw_args && last_arg_matches_any_keywords
+
+            # If we don't allow the keyword splat, and the last hash contains extra keys
+            # then it is a positional hash.
+            return false if !@allows_any_kw_args && has_extra_keys
```

Comment:
I guess there's a problem between 2.7 and 3.0+ with (`spec/rspec/support/method_signature_verifier_spec.rb:372`)
```ruby
> def arity_kw(x, y = {}, z:2); end

> arity_kw(nil, {a: 1}, {z: 2})
```
it works fine on 2.7 but fails on 3.0+:
```
ArgumentError: wrong number of arguments (given 3, expected 1..2)
```

To me, it seems that this expectation is not justified on Ruby 2.7:
```
expect(valid?(nil, :a => 1)).to eq(false)
```
as `y` is initialized with `{:a => 1}`.

I didn't look at other few failing examples yet.